process_report/process_report.py

```diff
@@ -112,21 +194,107 @@ def remove_billables(dataframe, pi, projects, output_file):
 
     So this *keeps* the projects/pis that should not be billed.
     """
-    filtered_dataframe = dataframe[dataframe['Manager (PI)'].isin(pi) | dataframe['Project - Allocation'].isin(projects)]
+    filtered_dataframe = dataframe[dataframe[PI_FIELD].isin(pi) | dataframe[PROJECT_FIELD].isin(projects)]
     filtered_dataframe.to_csv(output_file, index=False)
 
+
+def validate_billables(dataframe):
+    # Validate PI name
+    invalid_pi_projects = dataframe[pandas.isna(dataframe[PI_FIELD])]
+    for i, row in invalid_pi_projects.iterrows():
+        print(f'Warning: Project {row[PROJECT_FIELD]} has empty PI field')
+    dataframe = dataframe[~pandas.isna(dataframe[PI_FIELD])]
+
+    return dataframe
+
+
+def export_billables(dataframe, output_file):
+    dataframe.to_csv(output_file, index=False)
+
+
 def export_pi_billables(dataframe: pandas.DataFrame, output_folder):
     if not os.path.exists(output_folder):
         os.mkdir(output_folder)
 
-    invoice_month = dataframe['Invoice Month'].iat[0]
-    pi_list = dataframe['Manager (PI)'].unique()
+    invoice_month = dataframe[INVOICE_DATE_FIELD].iat[0]
+    pi_list = dataframe[PI_FIELD].unique()
 
     for pi in pi_list:
-        pi_projects = dataframe[dataframe['Manager (PI)'] == pi]
-        pi_instituition = pi_projects['Institution'].iat[0]
+        if pandas.isna(pi): 
+            continue
+        pi_projects = dataframe[dataframe[PI_FIELD] == pi]
+        pi_instituition = pi_projects[INSTITUTION_FIELD].iat[0]
         pi_projects.to_csv(output_folder + f"/{pi_instituition}_{pi}_{invoice_month}.csv")
-        
+
+
+def apply_credits_new_pi(dataframe, old_pi_file):
+    new_pi_credit_code = "0002"
+    new_pi_credit_amount = 1000
+
+    dataframe[CREDIT_FIELD] = None
+    dataframe[CREDIT_CODE_FIELD] = None
+    dataframe[BALANCE_FIELD] = 0
+
+    old_pi_dict = load_old_pis(old_pi_file)
+
+    current_pi_list = dataframe[PI_FIELD].unique()
+    invoice_month = dataframe[INVOICE_DATE_FIELD].iat[0]
+
+    for pi in current_pi_list:
+        pi_projects = dataframe[dataframe[PI_FIELD] == pi]
+
+        if is_old_pi(old_pi_dict, pi, invoice_month):
+            for i, row in pi_projects.iterrows():
+                dataframe.at[i, BALANCE_FIELD] = row[COST_FIELD]
+        else:
+            remaining_credit = new_pi_credit_amount
+            for i, row in pi_projects.iterrows():
+                project_cost = row[COST_FIELD]
+                applied_credit = min(project_cost, remaining_credit)
+
+                dataframe.at[i, CREDIT_FIELD] =  applied_credit
+                dataframe.at[i, CREDIT_CODE_FIELD] = new_pi_credit_code
+                dataframe.at[i, BALANCE_FIELD] = row[COST_FIELD] - applied_credit
+                remaining_credit -= applied_credit
+
+                if remaining_credit == 0:
```

Comment:
I figured the break statement would just help save a bit more time. We could break early instead of iterating through all the projects when all the credits have been spent.

Reply:
But we still want to populate those Credits fields with 0, no?

Reply:
From the [invoices](https://docs.google.com/spreadsheets/d/1O7Kqu0cF2nVqNTuNWdzl9NXX6Gg5eB4IZW4LQDItU3w/edit#gid=1189051626) @joachimweyl shared with me, its seems it's fine to leave the credit field empty if no credit is applied

Reply:
Alright.